libraries/chain/webassembly/eos-vm-oc/compile_monitor.cpp

```diff
@@ -0,0 +1,309 @@
+      auto& [code, socket] = *current_compile_it;
+      socket.async_wait(local::datagram_protocol::socket::wait_read, [this, current_compile_it](auto ec) {
+         //at this point we only expect 1 of 2 things to happen: we either get a reply (success), or we get no reply (failure)
+         auto& [code, socket] = *current_compile_it;
+         auto [success, message, fds] = read_message_with_fds(socket);
+         eosvmoc_message reply;
+         if(!success) {
+            reply = wasm_compilation_result_message{code, compilation_result_unknownfailure{}, _allocator->get_free_memory()};
+         }
+         else {
+            //XXX code_compilation_result_message, we should have 2 fds: code & initialmem
+            message.visit(overloaded {
+               [&, &fds=fds, &code=code](const code_compilation_result_message& result) {
+                  ///XXX verify there are 2 fds
+                  void* code_ptr = _allocator->allocate(get_size_of_fd(fds[0]));
+                  void* mem_ptr = _allocator->allocate(get_size_of_fd(fds[1]));
+                  if(code_ptr == nullptr || mem_ptr == nullptr) {
+                     _allocator->deallocate(code_ptr);
+                     _allocator->deallocate(mem_ptr);
+                     reply = wasm_compilation_result_message{code, compilation_result_toofull{}, _allocator->get_free_memory()};
+                     return;
+                  }
+                  ///XXX check if either is NULL and back out
+                  copy_memfd_contents_to_pointer(code_ptr, fds[0]);
+                  copy_memfd_contents_to_pointer(mem_ptr, fds[1]);
+
+                  reply = wasm_compilation_result_message{code, code_descriptor {
+                        code.code_id,
+                        code.vm_version,
+                        0,
+                        (uintptr_t)code_ptr - (uintptr_t)_code_mapping,
+                        result.start,
+                        result.apply_offset,
+                        result.starting_memory_pages,
+                        (uintptr_t)mem_ptr - (uintptr_t)_code_mapping,
+                        (unsigned)get_size_of_fd(fds[1]),
+                        result.initdata_prolouge_size
+                     },
+                     _allocator->get_free_memory()
+                  };
+               },
+               [&, &code=code](const auto&) {
+                  //XXX anything else is an error
+                  reply = wasm_compilation_result_message{code, compilation_result_unknownfailure{}, _allocator->get_free_memory()};
+               }
+            });
+         }
+
+         write_message_with_fds(_nodeos_instance_socket, reply);
+
+         //either way, we are done
+         _ctx.post([this, current_compile_it]() {
+            current_compiles.erase(current_compile_it);
+         });
+      });
+
+   }
+
+   boost::signals2::signal<void()> connection_dead_signal;
```

Comment:
A signal seems like overkill.  There's always exactly one.

Reply:
Nope, there are some unittests that create two controllers ergo two wasm_runtimes ergo two connections

Reply:
Sure there can be more than one controller, but the connection_dead_signal is only used to erase the compile_monitor_session that it is part of.  ``connect`` is only called once, right after the compile_monitor_session is emplaced into compile_monitor::_compile_sessions.

Reply:
ah okay, I dunno that whole thing is just a pattern I typically use for asio server + asio clients so that I avoid shared_ptrs and client objects are completely decoupled